external/fv3fit/fv3fit/reservoir/adapters.py

```diff
@@ -0,0 +1,169 @@
+@io.register("reservoir-adapter")
+class ReservoirDatasetAdapter(Predictor):
+    MODEL_DIR = "reservoir_model"
+
+    def __init__(
+        self,
+        model: ReservoirComputingModel,
+        input_variables: Iterable[Hashable],
+        output_variables: Iterable[Hashable],
+    ) -> None:
+        self.model = model
+        self.input_variables = input_variables
+        self.output_variables = output_variables
+        self.model_adapter = DatasetAdapter(
+            input_variables=input_variables,
+            output_variables=output_variables,
+            rank_dims=model.rank_divider.rank_dims,
+        )
+
+    def predict(self, inputs: xr.Dataset) -> xr.Dataset:
+        # inputs arg is not used, but is required by Predictor signature and prog run
+        prediction_arr = self.model.predict()
+        return self.model_adapter.output_array_to_ds(
+            prediction_arr, output_dims=list(inputs.dims)
+        )
+
+    def increment_state(self, inputs: xr.Dataset):
+        xy_input_arrs = self.model_adapter.input_dataset_to_arrays(
+            inputs, self.input_variables
+        )  # x, y, feature dims
+        self.model.increment_state(xy_input_arrs)
+
+    def reset_state(self):
+        self.model.reset_state()
+
+    def dump(self, path):
+        self.model.dump(os.path.join(path, self.MODEL_DIR))
+
+    @classmethod
+    def load(cls, path: str) -> "ReservoirDatasetAdapter":
+        model = ReservoirComputingModel.load(os.path.join(path, cls.MODEL_DIR))
+        model.reset_state()
+        adapter = cls(
+            input_variables=model.input_variables,
+            output_variables=model.output_variables,
+            model=model,
+        )
+        return adapter
+
+
+@io.register("hybrid-reservoir-adapter")
+class HybridReservoirDatasetAdapter(Predictor):
+    MODEL_DIR = "hybrid_reservoir_model"
+
+    def __init__(
+        self,
+        model: HybridReservoirComputingModel,
+        input_variables: Iterable[Hashable],
+        output_variables: Iterable[Hashable],
+    ) -> None:
+        self.model = model
+        self.input_variables = list(set(input_variables).union(model.hybrid_variables))
+        self.output_variables = output_variables
+        self.model_adapter = DatasetAdapter(
+            input_variables=self.input_variables,
+            output_variables=output_variables,
+            rank_dims=model.rank_divider.rank_dims,
+        )
+
+    def predict(self, inputs: xr.Dataset) -> xr.Dataset:
+        xy_input_arrs = self.model_adapter.input_dataset_to_arrays(
+            inputs, self.model.hybrid_variables
+        )  # x, y, feature dims
+
+        prediction_arr = self.model.predict(xy_input_arrs)
+        return self.model_adapter.output_array_to_ds(
+            prediction_arr, output_dims=list(inputs.dims)
+        )
+
+    def increment_state(self, inputs: xr.Dataset):
+        xy_input_arrs = self.model_adapter.input_dataset_to_arrays(
+            inputs, self.input_variables
```

Comment:
I think this should be `self.model.input_variables` since the `self.input_variables` includes the hybrids.